style/react/README.md

```diff
@@ -0,0 +1,50 @@
+# React
+
+ * Use React in [Strict Mode]
+ * Use React with [TypeScript](../typescript/README.md)
+ * Use [Jest](../testing-jest/README.md) for unit testing
+ * Use [styled-components] for CSS-in-JS
+ * Use [Apollo] with [code generation] when working with [GraphQL]
+ * Use [React Router] for routing in SPAs
+ * Avoid deeply nested routing when using React Router
+ * Prefer [Function Components] over [Class Components]
+ * Prefer keeping a single component in each file
+ * Use `PascalCase` for component names and their file names
+ * Use [React Hooks]
+ * Use pre-built hooks when possible (e.g. [streamich/react-use])
+ * Use [custom hooks] to encapsulate stateful logic outside a component
+ * Avoid [Forwarding Refs]
+ * Avoid [Higher-Order Components] and [recompose]
+ * Prefer the `children` prop over [render props]
+ * Prefer using [TypeScript prop interfaces] over [PropTypes]
+ * Prefer the [short syntax] when using [Fragments]
+ * Prefer [React Contexts] over [Redux]
+ * Prefer [Controlled Components] over [Uncontrolled Components]
```

Comment:
That is correct -- an uncontrolled component will allow the user to change its value directly, where a controlled component will always set and maintain the value that is passed as a prop to the element.

This was an attempt to point out that you almost always want a controlled component, because often you are driving a form element off component state: you have a state variable and you set it `onChange` of the element, and pass the current state as the `value` prop to the element.

An antipattern here would be to try and accomplish the same with an uncontrolled component: you have a state variable and you set it `onChange` of the element, but you **don't** pass the current state as the `value` prop to the element. This might coincidentally work if the form element doesn't get re-rendered, but breaks down once a re-render happens (because the element loses the current value).

But the more I think about it, this situation is super specific and probably not fit for a best-practice.